src/lib.rs

```diff
@@ -2103,6 +2102,68 @@ impl Url {
         Ok(())
     }
 
+    /// Return whether the URL is special.
+    ///
+    /// A URL is special if its scheme is one of the following:
+    ///
+    /// "http" | "https" | "ws" | "wss" | "ftp" | "gopher" | "file"
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use url::Url;
+    /// # use url::ParseError;
+    ///
+    /// # fn run() -> Result<(), ParseError> {
+    /// let url = Url::parse("ftp://rms@example.com")?;
+    /// assert!(url.is_special());
+    ///
+    /// let url = Url::parse("file://foo.bar")?;
+    /// assert!(url.is_special());
+    ///
+    /// let url = Url::parse("unix:/run/foo.socket")?;
+    /// assert!(!url.is_special());
+    ///
+    /// let url = Url::parse("data:text/plain,Stuff")?;
+    /// assert!(!url.is_special());
+    /// # Ok(())
+    /// # }
+    /// # run().unwrap();
+    /// ```
+    pub fn is_special(&self) -> bool {
+        SchemeType::from(self.scheme()).is_special()
+    }
+
+    /// Return whether the URL includes credentials.
+    ///
+    /// A URL includes credentials if its username or password is not the empty string.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use url::Url;
+    /// # use url::ParseError;
+    ///
+    /// # fn run() -> Result<(), ParseError> {
+    /// let url = Url::parse("https://username:password@www.my_site.com")?;
+    /// assert!(url.includes_credentials());
+    ///
+    /// let url = Url::parse("https://username@www.my_site.com")?;
+    /// assert!(url.includes_credentials());
+    ///
+    /// let url = Url::parse("https://www.my_site.com")?;
+    /// assert!(!url.includes_credentials());
+    ///
+    /// let url = Url::parse("https://@www.my_site.com")?;
+    /// assert!(!url.includes_credentials());
+    /// # Ok(())
+    /// # }
+    /// # run().unwrap();
+    /// ```
+    pub fn includes_credentials(&self) -> bool {
+        self.username() != "" || self.password().unwrap_or(&"") != ""
```

Comment:
I'm confused about this code, is it even possible to have a password without username?